pkg/raft/raft.go

```diff
@@ -1329,7 +1338,46 @@ func (r *raft) becomeLeader() {
 		panic("invalid transition [follower -> leader]")
 	}
 	r.step = stepLeader
-	r.reset(r.Term)
+
+	// A special visit function to use the recorded hints and optimize the first
+	// round of MsgApp after becoming leader.
+	specialVisit := func(id pb.PeerID, pr *tracker.Progress) {
+		// Like regular reset, next is set to r.raftLog.lastIndex() + 1 on default.
+		next := r.raftLog.lastIndex() + 1
+		// ConflictIdx is either a best guess on where the voter's log matches the
+		// leader's, or r.raftLog.lastIndex()
+		conflictIdx, isMatch, ok := r.electionTracker.GetGuessMatchEntryID(id)
+		// A conflictIdx is recorded and the voter voted during election.
+		if ok && r.electionTracker.GetVote(id) {
+			next = uint64(conflictIdx) + 1
+		}
+
+		*pr = tracker.Progress{
+			Match:       0,
+			MatchCommit: 0,
+			Next:        next,
+			Inflights:   tracker.NewInflights(r.maxInflight, r.maxInflightBytes),
+			IsLearner:   pr.IsLearner,
+		}
+		if id == r.id {
+			pr.Match = r.raftLog.lastIndex()
+		}
+
+		if isMatch {
+			// Follower's log matches the leader's log at next-1.
+			// Set the follower's state to StateReplicate early.
+			r.becomeReplicate(pr)
+			// NB: this is a bit of a hack, but we need to set the Next again since
+			// r.becomeReplicate sets pr.Next to match + 1.
+			// Where match is 0 in this case.
+			// TODO(pav-kv): consider setting match to (next - 1) in this case to
+			// avoid this explicit assignment.
+			pr.Next = next
+		}
+	}
+
+	r.reset(r.Term, specialVisit)
```

Comment:
Perhaps the `r.trk.Visit` below can be merged into this visitor. The `TODO(pav-kv)` a few lines below already encompasses this, so let's do in a follow-up.